perfkitbenchmarker/linux_benchmarks/iperf_benchmark.py

```diff
@@ -98,63 +126,191 @@ def _RunIperf(sending_vm, receiving_vm, receiving_ip_address, thread_count,
-    # which happens when threads start at different times.  The code
-    # below will tend to overestimate a bit.
-    thread_values = re.findall(r'\[.*\d+\].*\s+(\d+\.?\d*).Mbits/sec', stdout)
-
-    if len(thread_values) != thread_count:
-      raise ValueError('Only %s out of %s iperf threads reported a'
-                       ' throughput value.' %
-                       (len(thread_values), thread_count))
-
-  total_throughput = 0.0
-  for value in thread_values:
-    total_throughput += float(value)
-
-  metadata = {
-      # The meta data defining the environment
-      'receiving_machine_type': receiving_vm.machine_type,
-      'receiving_zone': receiving_vm.zone,
-      'sending_machine_type': sending_vm.machine_type,
-      'sending_thread_count': thread_count,
-      'sending_zone': sending_vm.zone,
-      'runtime_in_seconds': FLAGS.iperf_runtime_in_seconds,
-      'ip_type': ip_type
-  }
-  return sample.Sample('Throughput', total_throughput, 'Mbits/sec', metadata)
+
+  if protocol == TCP:
+
+    iperf_cmd = ( f'iperf -e --client {receiving_ip_address} --port {IPERF_PORT} '
+                  f'--format m --time {FLAGS.iperf_runtime_in_seconds} -P {thread_count}')
+
+    if FLAGS.iperf_tcp_per_stream_bandwidth:
+      iperf_cmd += f' -b {FLAGS.iperf_tcp_per_stream_bandwidth}M'
+
+    # the additional time on top of the iperf runtime is to account for the
+    # time it takes for the iperf process to start and exit
+    timeout_buffer = FLAGS.iperf_timeout or 30 + thread_count
+    stdout, _ = sending_vm.RemoteCommand(iperf_cmd, should_log=True,
+                                         timeout=FLAGS.iperf_runtime_in_seconds + timeout_buffer)
+
+    multi_thread = re.search((r'\[SUM\]\s+\d+\.\d+-\d+\.\d+\s\w+\s+(?P<transfer>\d+)\s\w+\s+(?P<throughput>\d+)'
+                              r'\s\w+\/\w+\s+(?P<write>\d+)\/(?P<err>\d+)\s+(?P<retry>\d+)\s*'),
+                             stdout)
+
+    window_size_match = re.search(r'TCP window size: (?P<size>\d+\.?\d+) (?P<units>\S+)', stdout) 
+    window_size = float(window_size_match.group('size'))
+    window_size_unit = window_size_match.group('units')
+
+    buffer_size = float(re.search(r'Write buffer size: (?P<buffer_size>\d+\.\d+) \S+', stdout).group('buffer_size'))
+
+    # Iperf output is formatted differently when running with multiple threads vs a single thread
+    if multi_thread:
+      # Write, error, retry
+      write = int(multi_thread.group('write'))
+      err = int(multi_thread.group('err'))
+      retry = int(multi_thread.group('retry'))
+
+    # if single thread
+    else:
+      # Write, error, retry
+      match = re.search(r'\d+ Mbits\/sec\s+(?P<write>\d+)\/(?P<err>\d+)\s+(?P<retry>\d+)', stdout)
+      write = int(match.group('write'))
+      err = int(match.group('err'))
+      retry = int(match.group('retry'))
+
+    match = re.findall(r'\d+ Mbits\/sec\s+ \d+\/\d+\s+\d+\s+(-*\d+)(\w+)\/(\d+)\s+(\w+)\s+(\d+\.\d+)', stdout)
```

Comment:
Can you name the capture groups here as well?

Reply:
Named the capture groups. It improves readability, but .findall() ignores the names.